### PR TITLE
feat: weighted load-balancing via `package.json` config

### LINKED ISSUE #13: Load balance with weighted round robin to give partitions approximately equal runtime
I'm looking at having a couple dozen packages managed through lerna, and due to some name prefixing conventions, it turns out that when putting all the packages in alphabetical order, the first half have vastly shorter test runtime than the later half.  

So, splitting these into 4 chunks results in a pretty imbalanced runtimes between partitions.

Here, I'm proposing a greedy, weighted load-balancing algorithm as an option to balance partitions more evenly.  The weights can be provided via a package.json property and can be accessed through the lerna `Package` class's `.get()` method.  The name of the property can be a command line param which defaults to `lernaPartitionTestWeight`.

### PR DESCRIPTION
**Pull request by kdnutrien**
### Add load-balance feature to partition by hinted weight for more equal runtimes

This PR closes #13 by adding a greedy round-robin weighted load balancing algorithm.  Each package can declare a numeric weight in `package.json`, and with a `--load-balance` CLI argument, the packages will be grouped in partitions with approximately equal runtimes.

Tests are added as well as README usage notes.

Use of the `--packageTestWeightKey ` is likely rare, but for advanced use cases with multiple CI processes (say, typechecking, tests, and linting) it is possible that different partitioning strategies may be desired for each of these.

### CODE REVIEW

**Review by jeffatnutrien — APPROVED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -93,6 +93,34 @@ others, it does, specifically:
...
+To address this, `lerna-parallelism` offers a load balancing mode which uses a
+weighted round-robin algorithm, informed by a weight "hint" specified in each 
+package's `package.json`.
+
+This adds two additional CLI options in addition to those above:
+
+- `--loadBalance`: Toggles load-balancing mode. Defaults to
+  `false`.
+- `--packageTestWeightKey myProjectWeight`: The lookup key used to read the project's
+  weight from its package.json. Defaults to `lernaPackageTestWeight`.
```
Comment (buschtoens) [inline README.md:112]: Is there a reason you chose `--packageTestWeightKey` over just `--packageWeightKey`?

Would you be okay with changing this?

```suggestion
- `--packageWeightKey myProjectWeight`: The lookup key used to read the project's
  weight from its `package.json`. Defaults to `lernaPackageWeight`.
```

File: README.md [hunk 2]
```diff
@@ -93,6 +93,34 @@ others, it does, specifically:
...
+
+### Usage
+
+`lerna-parallelism --loadBalance run test`
+
+And add a `lernaPackageTestWeight` property (numeric) to your package.json for each package.
+If it is missing, `1` is the default weight assigned.
+
+If you have a need to partition differently for multiple CI tasks, you can use
+`--packageTestWeightKey` to specify which weight property should be read from package.json.
```
Comment (buschtoens) [inline README.md:122]:
```suggestion
If you have a need to partition differently for multiple CI tasks, you can use
`--packageWeightKey` to specify which weight property should be read from `package.json`.
```

File: commands/run.js [hunk 3]
```diff
@@ -1,19 +1,32 @@
...
+
+      const { packages, logMessage } = loadBalance
+        ? getLoadBalancedPackages(
+            this.packagesWithScript,
+            split,
+            partition,
+            packageTestWeightKey,
+            this.logger
+          )
+        : getSplitPackages(this.packagesWithScript, split, partition);
```
Comment (buschtoens) [inline commands/run.js:29]:
```suggestion
        partition,
        loadBalance,
        packageWeightKey
      } = this.options;

      const { packages, logMessage } = loadBalance
        ? getLoadBalancedPackages(
            this.packagesWithScript,
            split,
            partition,
            packageWeightKey,
            this.logger
          )
        : getSplitPackages(this.packagesWithScript, split, partition);
```

File: options/split.js [hunk 4]
```diff
@@ -22,12 +22,25 @@ module.exports.splittable = yargs =>
...
+      type: 'boolean',
+      default: false,
+      defaultDescription:
+        'Use a greedy round-robin load balancing algo instead of default split'
+    })
+    .option('packageTestWeightKey', {
+      type: 'string',
+      default: 'lernaPackageTestWeight',
+      defaultDescription:
+        "The lookup key to use for reading the project's weight from its package.json"
```
Comment (buschtoens) [inline options/split.js:43]:
```suggestion
    .option('packageWeightKey', {
      type: 'string',
      default: 'lernaPackageWeight',
      defaultDescription:
        "The lookup key to use for reading the project's weight from its package.json"
```

File: options/split.js [hunk 5]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
     packages: chunk
   };
 };
+
+module.exports.getLoadBalancedPackages = function (
+  packages,
+  split,
+  partition,
+  packageTestWeightKey,
```
Comment (buschtoens) [inline options/split.js:97]:
```suggestion
  packageWeightKey,
```

File: options/split.js [hunk 6]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+  const partitionWeights = {};
+  partitions.forEach((_, index) => {
+    partitionWeights[index] = 0;
+  });
+
+  // Sort packages by weight in descending order.
+  // We will put the largest remaining weighted package in the partition with the least total weight.
+  const packagesSortedByWeightDesc = [...packages].sort(
+    (a, b) =>
+      (b.get(packageTestWeightKey) || 1) - (a.get(packageTestWeightKey) || 1)
```
Comment (buschtoens) [inline options/split.js:120]:
```suggestion
      (b.get(packageWeightKey) || 1) - (a.get(packageWeightKey) || 1)
```

File: options/split.js [hunk 7]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+    }
+
+    return minId;
+  };
+
+  packagesSortedByWeightDesc.forEach(project => {
+    const lightestPartition = getLightestPartition();
+    logger.debug(
+      `Adding package ${project.name} (weight: ${project.get(
+        packageTestWeightKey
```
Comment (buschtoens) [inline options/split.js:140]:
```suggestion
        packageWeightKey
```

File: options/split.js [hunk 8]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+      )}) to partition ${lightestPartition}`
+    );
+    partitions[lightestPartition].push(project);
+    partitionWeights[lightestPartition] +=
+      project.get(packageTestWeightKey) || 1;
```
Comment (buschtoens) [inline options/split.js:145]:
```suggestion
      project.get(packageWeightKey) || 1;
```

File: options/split.js [hunk 9]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+  });
+
+  logger.debug(
+    partitions.map(part =>
+      part.map(pack => [pack.name, pack.get(packageTestWeightKey) || 1])
```
Comment (buschtoens) [inline options/split.js:150]:
```suggestion
      part.map(pack => [pack.name, pack.get(packageWeightKey) || 1])
```

**Review by buschtoens — APPROVED** [review]
Thank you for this awesome addition!

I'm sorry, that I'm only getting to this now. The notification must have slipped my attention.

There's nothing blocking the merge, but I'd like to change `--packageTestWeightKey` to `--packageWeightKey` before releasing this.

File: options/split.js [hunk 10]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+  const partitions = [];
+  for (let i = 0; i < split; i++) {
+    partitions.push([]);
+  }
+
+  // Init partition weight totals
+  const partitionWeights = {};
+  partitions.forEach((_, index) => {
+    partitionWeights[index] = 0;
+  });
```
Comment (buschtoens) [inline options/split.js:114]:
```suggestion
  // Init partition weight totals
  const partitionWeights = Array.from({ length: partitions.length }).fill(0);
```

File: options/split.js [hunk 11]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+
+  // Sort packages by weight in descending order.
+  // We will put the largest remaining weighted package in the partition with the least total weight.
+  const packagesSortedByWeightDesc = [...packages].sort(
+    (a, b) =>
+      (b.get(packageWeightKey) || 1) - (a.get(packageWeightKey) || 1)
+  );
```
Comment (buschtoens) [inline options/split.js:121]:
```suggestion
  // Sort packages by weight in descending order.
  // We will put the largest remaining weighted package in the partition with the least total weight.
  const packagesSortedByWeightDesc = [...packages].sort(
    (a, b) => (b.get(packageWeightKey) || 1) - (a.get(packageWeightKey) || 1)
  );
```

File: options/split.js [hunk 12]
```diff
@@ -76,3 +89,79 @@ module.exports.getSplitPackages = function (packages, split, partition) {
...
+    const lightestPartition = getLightestPartition();
+    logger.debug(
+      `Adding package ${project.name} (weight: ${project.get(
+        packageWeightKey
+      )}) to partition ${lightestPartition}`
+    );
+    partitions[lightestPartition].push(project);
+    partitionWeights[lightestPartition] +=
+      project.get(packageWeightKey) || 1;
+  });
```
Comment (buschtoens) [inline options/split.js:146]:
```suggestion
  for (const project of packagesSortedByWeightDesc) {
    const lightestPartition = getLightestPartition();
    logger.debug(
      `Adding package ${project.name} (weight: ${project.get(
        packageWeightKey
      )}) to partition ${lightestPartition}`
    );
    partitions[lightestPartition].push(project);
    partitionWeights[lightestPartition] += project.get(packageWeightKey) || 1;
  }
```

**Comment by buschtoens** [comment]
@kdnutrien Your feature is now released as https://github.com/ClarkSource/lerna-parallelism/releases/tag/v0.2.1! 🚀
<sup>(Technically already https://github.com/ClarkSource/lerna-parallelism/releases/tag/v0.2.0, but I messed up. 🙈)</sup>
The docs are here: [`--distribute-by weight`](https://github.com/ClarkSource/lerna-parallelism#--distribute-by-weight) 📖

Thank you for your idea and this great contribution!

FYI the main changes I made are:

- Rename `--loadBalance` to `--distribute-by weight`. Defaults to `--distribute-by count`.
- Rename `--packageTestWeightKey` to `--packageWeightKey`.
- Change default of `--packageWeightKey` from `lernaPackageTestWeight` to `lernaPackageWeight`.

**Comment by kdnutrien** [comment]
Awesome, thank you!

**Pull request merged**